core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/CBuildConfiguration.java

```diff
@@ -128,11 +128,12 @@ protected CBuildConfiguration(IBuildConfiguration config, String name) throws Co
 		String typeId = settings.get(TOOLCHAIN_TYPE, ""); //$NON-NLS-1$
 		String id = settings.get(TOOLCHAIN_ID, ""); //$NON-NLS-1$
 		IToolChainManager toolChainManager = CCorePlugin.getService(IToolChainManager.class);
+		// FIXME: typeId and id are always empty string, and therefore tc is always null.
```

Comment:
I still don't understand this comment. I can see typeId and id being non-empty string without your change applied if I do this:

1. Create CMake project
2. Build project
3. Restart Eclipse

At this point the typeId/id will be non-empty.

To me this indicates that the bug fix has missed a use case that may or may not be broken with this PR.

>  It is fixed to the "run" launch mode, but the "run" build configs are not created with this.

I don't think this statement is true either. Using the above flow launchMode == debug

Reply:
No problem.
I'm mainly focused on Makefile flow. I did not look at Autotools and Meson. I never used it. I can have look.
I may have overlooked a use case. I will look again.

Reply:
I quickly tried Autotools and Meson.
For Meson the problem is similar as in Makefile and CMake.
For Autotools you can get the problem even without closing the project. First compile for Run end then for Debug gives the problem. If you compile first for Debug and then for Run it compiles OK. After closing and opening the project you get the same problem as with the others.

Reply:
> I still don't understand this comment. I can see typeId and id being non-empty string without your change applied if I do this:
> 
> 1. Create CMake project
> 2. Build project
> 3. Restart Eclipse
> 
> At this point the typeId/id will be non-empty.

You are right. When you restart Eclipse it works as expected. The problem occurs when you close the project and reopen the project. You don't need to exit Eclipse. When you reopen the project the settings have no properties. This is the root of the problem. Then typeId and id will be empty string. tc will be null, and an exception is thrown. And then the CBuildConfigurationManager puts the configuration in the noConfigs list. After that it will not be possible to build the project for debug.

I don't know (yet) why the settings properties are lost when the project is closed. My fix is to avoid using this constructor, because it also not used for the Run launch mode. Fixing the loss of the settings could be a better fix. But why was it not used for Run mode?

Reply:
I have changed the patch to keep the build configuration's settings at project closure. This fixes the problem for all Core Build project types. I added some comment to CBuildConfiguration to make it a bit easier to understand.